File: elixir/protein-translation/lib/protein_translation.ex
defmodule ProteinTranslation do
  @doc """
  Given an RNA string, return a list of proteins specified by codons, in order.
  """
  @spec of_rna(String.t()) :: {:ok, list(String.t())} | {:error, String.t()}
  def of_rna(rna) do
  end

  @doc """
  Given a codon, return the corresponding protein

  UGU -> Cysteine
  UGC -> Cysteine
  UUA -> Leucine
  UUG -> Leucine
  AUG -> Methionine
  UUU -> Phenylalanine
  UUC -> Phenylalanine
  UCU -> Serine
  UCC -> Serine
  UCA -> Serine
  UCG -> Serine
  UGG -> Tryptophan
  UAU -> Tyrosine
  UAC -> Tyrosine
  UAA -> STOP
  UAG -> STOP
  UGA -> STOP
  """
  @spec of_codon(String.t()) :: {:ok, String.t()} | {:error, String.t()}
  def of_codon("AUG"), do: {:ok, "Methionine"}
  def of_codon(codon) when codon in ["UUU", "UUC"] do {:ok, "Phenylalanine"} end
  def of_codon(codon) when codon in ["UUA", "UUG"] do {:ok, "Leucine"} end
  def of_codon(codon) when codon in ["UCU", "UCC", "UCA", "UCG"] do {:ok, "Serine"} end
  def of_codon(codon) when codon in ["UCU", "UCC", "UCA", "UCG"] do {:ok, "Serine"} end
  def of_codon(codon) when codon in ["UAU", "UAC"] do {:ok, "Tyrosine"} end
  def of_codon(codon) when codon in ["UGU", "UGC"] do {:ok, "Cysteine"} end
  def of_codon(codon) when codon in ["UGG"] do {:ok, "Tryptophan"} end
  def of_codon(codon) when codon in ["UAA", "UAG", "UGA"] do {:ok, "STOP"} end
  def of_codon(codon) do
  end
end

File: fsharp/protein-translation/ProteinTranslation.fs
module ProteinTranslation

let proteins (rna: string): string list =
    let codonToProtein (rna: string): (string * string) option =
        match rna[0..2] with
        | "AUG" -> Some ("Methionine",  rna[3..])
        | "UAC" | "UAU" -> Some ("Tyrosine", rna[3..])
        | "UCA" | "UCC" | "UCG" | "UCU" -> Some ("Serine", rna[3..])
        | "UGC" | "UGU" -> Some ("Cysteine", rna[3..])
        | "UGG" -> Some("Tryptophan", rna[3..])
        | "UUA" | "UUG" -> Some ("Leucine", rna[3..])
        | "UUU" | "UUC" -> Some ("Phenylalanine", rna[3..])
        | "UAA" | "UAG" | "UGA" | "" -> None
        | _ -> failwith "Unknown coding"
        
    List.unfold codonToProtein rna